### DISCUSSION TITLE
Handling Interaction Payloads

### DISCUSSION THREAD

**Discussion by svisconti02** [stoiveyp/Slack.NetStandard#85] [Q&A]
Hello, 

I'm calling a method ProcessMessages() based on the SocketSample, which will DM a set of interactive blocks to a user upon recieving a slash command. This works as expected; however, I'm failing to receive any sort of envelope when interacting with the blocks. Interactivity is enabled on the application, and I have been able to receive slash commands as well, so I'm wondering what is going wrong. I appreciate any help, thanks!

     
 private static async Task ProcessMessages(SocketModeClient client, CancellationToken token)
        {
            await foreach (var envelope in client.EnvelopeAsyncEnumerable(token))
            {

                Console.WriteLine(envelope.Payload.ToString());  //trying to hit here for interaction

                if (envelope.Payload is SlashCommand) {} //sends some interaction blocks

**Comment by stoiveyp** [comment]
Morning @svisconti02! Thanks for reaching out

I've just cloned the repo from scratch so I could ensure I got the same experience as someone using the library. I'm probably going to repeat some of what you've told me - but it's just to make sure I've not misunderstood anything.

So I have an app with socket mode enabled. In the socket mode settings I see this screen, so as you've said - slash commands and interaction are both enabled.

![image](https://github.com/stoiveyp/Slack.NetStandard/assets/147125/e330878b-baee-4aea-b998-8a71016ff216)

Then I've altered the ProcessMessages loop so that the slash command I've got returns (in my case) an action block with a button (I tweaked it a little so my console dumped out the block list I was sending)

```csharp
if (envelope.Payload is SlashCommand)
                {
                    var message = JsonConvert.SerializeObject(new Acknowledge
                    {
                        EnvelopeId = envelope.EnvelopeId,
                        Payload = new Message
                        {
                            Blocks = new List<IMessageBlock>
                            {
                                new Section("This is from a socket....shh!"),
                                new Actions
                                {
                                    BlockId = "testAction", Elements = new List<IMessageElement>
                                    {
                                        new Button { ActionId = "Test",Text="Test" }
                                    }
                                }
                            }
                        }
                    });
                    Console.WriteLine(message);
                    await client.Send(message, token);
                }
```

Then, I think the bit that you're asking for, is I extended the if statement to handle interaction coming from the blocks

```csharp
                else if (envelope.Payload is BlockActionsPayload)
                {

                }
```

When I click on the button (although it doesn't respond to anything) it will hit a breakpoint in this else statement when someone interacts with the blocks.

You could also check the envelope type, as this is what I do in the `EnvelopeConverter` to correct deserialize the envelope, but type checking gives you way more options as an app gets more complicated.

```csharp
else if (envelope.Type == "interactive")
```

Hope this helps - please let me know!

**Reply by svisconti02** [reply]
Thanks for the help! I replicated those steps and successfully hit the button and received a payload; however, I think my issue is a bit different. 

Currently, I send the block action as a DM via the http conversations.open and post.Message requests once I receive the slash command. Is it still possible to receive a payload from one of those interactions through the socket, or am I only allowed to do so if I send the ephemeral message? Appreciate any help as slack's documentation seems unclear. Thanks!

**Comment by stoiveyp** [comment]
Hi @svisconti02 - it doesn't matter how the blocks get sent, all slash commands, interactive elements and events will be sent through that loop if they're enabled in the socket mode settings 👍
